### PR TITLE
fix #1489

### LINKED ISSUE #1489: [BUG] incidence matrix not working ?
**Description of bug**

Hi there,
I was trying to solve a min cost flow problem using the incidence matrix of a `SimpleDiGraph`, and I got the impression that the function `incidence_matrix` is incorrect.

**How to reproduce**

```julia
julia> using LightGraphs

julia> function my_incidence_matrix(G)
           M = zeros(Int, nv(G), ne(G))
           for (k, e) in enumerate(edges(G))
               M[src(e), k] = -1
               M[dst(e), k] = +1
           end
           return M
       end
my_incidence_matrix (generic function with 1 method)

julia> G = complete_digraph(3)
{3, 6} directed simple Int64 graph

julia> M1 = Matrix(incidence_matrix(G))
3×6 Array{Int64,2}:
 -1  -1   1  -1   1  -1
  1   0  -1   0   0   1
  0   1   0   1  -1   0

julia> M2 = my_incidence_matrix(G)
3×6 Array{Int64,2}:
 -1  -1   1   0   1   0
  1   0  -1  -1   0   1
  0   1   0   1  -1  -1

julia> M1 .== M2
3×6 BitArray{2}:
 1  1  1  0  1  0
 1  1  1  0  1  1
 1  1  1  1  1  0
```

**Expected behavior**

Both "incidence matrices" (the one I computed and the one computed by `LightGraphs`) should be equal.

**Actual behavior**

They are different, in particular the one computed by `LightGraphs` has no zeros on the first row, even though vertex 1 is clearly not incident to all edges.

**Code demonstrating bug**

See above

**Version information**
 
`Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 6`
 
`Status `~/.julia/environments/v1.5/Project.toml`
  [093fc24a] LightGraphs v1.3.3`

**Additional context**

I think the bug may be due to lines 170-174 of `spectral.jl`, I don't understand their purpose and commenting them solves it.

### PR DESCRIPTION
**Pull request by gdalle**
Hi!
I modified the code in `spectral.jl` to make it correct. it became shorter and clearer in the process, using the `sparse(I, J, V)` constructor for the adjacency matrix instead of a heavier `SparseMatrixCSC` routine.
However I didn't know how to run the tests using the new version of `LightGraphs.jl`, hopefully this didn't break anything.

### CODE REVIEW

File: src/linalg/spectral.jl [hunk 1]
```diff
@@ -153,34 +153,14 @@ between `v` and `i`.
...
-        end
-    end
-
-    spmx = SparseMatrixCSC(n_v, n_e, colpt, rowval, nzval)
-    return spmx
+    I = vcat(src.(edges(g)), dst.(edges(g)))
+    J = vcat(1:n_e, 1:n_e)
+    V = vcat(
+        fill((isdir || oriented) ? -one(T) : one(T), n_e),
+        fill(one(T), n_e)
```
Comment (matbesancon) [inline src/linalg/spectral.jl:161]:
```suggestion
        fill(one(T), n_e),
```

trailing coma makes diff smaller

File: src/linalg/spectral.jl [hunk 2]
```diff
@@ -153,34 +153,14 @@ between `v` and `i`.
...
-            end
-        end
-    end
-
-    spmx = SparseMatrixCSC(n_v, n_e, colpt, rowval, nzval)
-    return spmx
+    I = vcat(src.(edges(g)), dst.(edges(g)))
+    J = vcat(1:n_e, 1:n_e)
+    V = vcat(
+        fill((isdir || oriented) ? -one(T) : one(T), n_e),
```
Comment (matbesancon) [inline src/linalg/spectral.jl:160]:
```suggestion
    (isdir || oriented) ? -fill(one(T), n) : fill(one(T), n),
```

this checks the condition once instead of each iteration (even if we can assume the compiler could do it)

File: src/linalg/spectral.jl [hunk 3]
```diff
@@ -153,34 +153,14 @@ between `v` and `i`.
...
-            end
-        end
-    end
-
-    spmx = SparseMatrixCSC(n_v, n_e, colpt, rowval, nzval)
-    return spmx
+    I = vcat(src.(edges(g)), dst.(edges(g)))
+    J = vcat(1:n_e, 1:n_e)
+    V = vcat(
+    (isdir || oriented) ? -fill(one(T), n) : fill(one(T), n),
```
Comment (matbesancon) [inline src/linalg/spectral.jl:160]:
```suggestion
    (isdir || oriented) ? -fill(one(T), n_e) : fill(one(T), n_e),
```
sorry, mistake on my end

**Comment by matbesancon** [comment]
To confirm, the tests also fail on Julia nightly on master, so unrelated

**Review by matbesancon — APPROVED** [review]

File: src/linalg/spectral.jl [hunk 4]
```diff
@@ -153,34 +153,14 @@ between `v` and `i`.
...
-
-    spmx = SparseMatrixCSC(n_v, n_e, colpt, rowval, nzval)
-    return spmx
+    I = vcat(src.(edges(g)), dst.(edges(g)))
+    J = vcat(1:n_e, 1:n_e)
+    V = vcat(
+    (isdir || oriented) ? -fill(one(T), n_e) : fill(one(T), n_e),
+        fill(one(T), n_e),
+    )
+    return sparse(I, J, V)
```
Comment (simonschoelly) [inline src/linalg/spectral.jl:163]:
```suggestion
    return sparse(I, J, V, nv(g), ne(g))
```
otherwise isolated vertices might be ignored.

**Comment by simonschoelly** [comment]
This is the behavior for undirected graphs with self-loops, that might be what we want, but is currently not reflected by the documentation:

```julia
julia> incidence_matrix(g)
3×2 SparseMatrixCSC{Int64, Int64} with 3 stored entries:
 1  ⋅
 1  ⋅
 ⋅  2

julia> incidence_matrix(g, Bool)
3×2 SparseMatrixCSC{Bool, Int64} with 3 stored entries:
 1  ⋅
 1  ⋅
 ⋅  1

julia> incidence_matrix(g, oriented=true)
3×2 SparseMatrixCSC{Int64, Int64} with 3 stored entries:
 -1  ⋅
  1  ⋅
  ⋅  0
  ```

**Comment by simonschoelly** [comment]
Maybe it would also be worth discussing, if this function should not also unoriented adjacency matrix for directed graphs?

**Comment by sbromberger** [comment]
This is great; thanks. My preference would be to merge this (with the isolated vertices tests that @matbesancon suggested) but then find a way to use the native constructor for a sparse matrix to avoid the allocations.

**Comment by gdalle** [comment]
> This is great; thanks. My preference would be to merge this (with the isolated vertices tests that @matbesancon suggested) but then find a way to use the native constructor for a sparse matrix to avoid the allocations.

Hi @sbromberger, do you mean the arrays I used to construct `I`, `J` and `V` before calling `sparse`? Would it be okay if those were iterators instead? Or do you prefer we used the explicit `SparseMatrixCSC` constructor ourselves? I used `sparse` for convenience and simplicity but I'm sure there is also a way to correct the initial (faulty) code that used `SparseMatrixCSC`.

**Comment by gdalle** [comment]
My fear is that since in our case using `sparse` is much easier, what we might end up doing if we use `SparseMatrixCSC` instead is effectively recoding `sparse` but less efficiently

**Comment by matbesancon** [comment]
Looking at what is required to build a sparse matrix, I agree that using `sparse` seems to be the recommended way. I don't think this method will be in hot loops much? Otherwise people would generate the matrix once and then use it in all iterations

**Comment by sbromberger** [comment]
@matbesancon I would like to move away from `sparse` into the native (inner?) constructor. The reason is that we're allocating `I`, `J`, and `V` and then dumping them into new allocations - I seem to recall this takes a long time for large matrices.

That shouldn't hold up merging this PR though - correctness is more important than performance - but it is an area we should look to optimize at some point soon.

**Comment by sbromberger** [comment]
One thing I'd love to see (in addition to the edge-case tests) is some benchmark results - even though the current way is incorrect, it is building a sparse matrix natively, so we should be able to see large differences if they exist.

File: src/linalg/spectral.jl (hunk 4 above)
Comment (sbromberger) [reply src/linalg/spectral.jl:163]: We should absolutely have a test for this.

**Comment by sbromberger** [comment]
Adding (since I'm on a 15-minute break from all-day meetings): if we decide to change the functionality of `incidence_matrix`, we should do that in a separate PR as it represents a breaking change.

**Comment by gdalle** [comment]
> @matbesancon I would like to move away from `sparse` into the native (inner?) constructor. The reason is that we're allocating `I`, `J`, and `V` and then dumping them into new allocations - I seem to recall this takes a long time for large matrices.
> 
> That shouldn't hold up merging this PR though - correctness is more important than performance - but it is an area we should look to optimize at some point soon.

That is why I was proposing to put them in lazy generators instead of arrays, that way they are only evaluated when dumped into the native sparse matrix format. A dumping which - I presume - the function `sparse` can do much better than us?

**Comment by gdalle** [comment]
> Adding (since I'm on a 15-minute break from all-day meetings): if we decide to change the functionality of `incidence_matrix`, we should do that in a separate PR as it represents a breaking change.

If you're referring to the case of graphs with self-loops, I must say I didn't worry about that when submitting the PR, the function was clearly wrong in general so I just wanted to make it right for my purposes. What should be the correct behavior? The one in the doc for now?

**Comment by sbromberger** [comment]
> That is why I was proposing to put them in lazy generators instead of arrays

TANSTAAFL. You're trading memory for CPU. Again, I'd be interested in the benchmarks. But in any case, `sparse()` calls `length()` on `I`, `J`, and `V`, so you're likely going to have to collect (i.e., allocate) the input vectors in their entirety anyway. And it might not even work since you need to ensure whatever generator you're passing conforms to `AbstractVector`.

> If you're referring to the case of graphs with self-loops

Sorry for the confusion; I wasn't. I was referring to @simonschoelly's comments [here](https://github.com/JuliaGraphs/LightGraphs.jl/pull/1524#issuecomment-775443570) and [here](https://github.com/JuliaGraphs/LightGraphs.jl/pull/1524#issuecomment-775452779). I think they're good ideas worthy of (and requiring) their own discussion and PR.

**Comment by matbesancon** [comment]
@gdalle regarding the test suggestion, a good simple test for it would be:
create a complete graph with 4 nodes, then add 3 vertices, then compute the incidence matrix and verify its size

**Comment by gdalle** [comment]
OK, I'm not sure I know how to amend a PR / add tests but I can look up both of these things, make the necessary adjustments and get back to you

**Comment by sbromberger** [comment]
> OK, I'm not sure I know how to amend a PR / add tests

It's easy. In your branch, just make the necessary changes, do a `git add -A`, `git commit -m <message>`, and then push the changes. This PR will automatically update.

**Comment by sbromberger** [comment]
@gdalle - just wanted to let you know I'm still on top of this. Once the tests get added I'm ready to do the final review. Thanks!

**Comment by gdalle** [comment]
Thanks! I had a quick look today but I don't know how to run my new tests with `Pkg`, since `Pkg.test("LightGraphs")` seems to test the standard version that I installed earlier. A quick Google search didn t help: should I just run the tests manually by executing the appropriate file? 

De: "Seth Bromberger" <notifications@github.com> 
À: "JuliaGraphs/LightGraphs.jl" <LightGraphs.jl@noreply.github.com> 
Cc: "Giom" <guillaume.dalle@enpc.fr>, "Mention" <mention@noreply.github.com> 
Envoyé: Samedi, Février 13, 2021 08:58:55 PM 
Objet: Re: [JuliaGraphs/LightGraphs.jl] fix #1489 (#1524) 

@gdalle - just wanted to let you know I'm still on top of this. Once the tests get added I'm ready to do the final review. Thanks!

**Comment by sbromberger** [comment]
Just add the tests and push them - let GitHub do the work for you :)

**Comment by gdalle** [comment]
yeah I know about CI but I m the kind of guy who wants to test the tests first ^^ I ll just create a Julia environment with my local version of LightGraphs 

De: "Seth Bromberger" <notifications@github.com> 
À: "JuliaGraphs/LightGraphs.jl" <LightGraphs.jl@noreply.github.com> 
Cc: "Giom" <guillaume.dalle@enpc.fr>, "Mention" <mention@noreply.github.com> 
Envoyé: Samedi, Février 13, 2021 10:42:46 PM 
Objet: Re: [JuliaGraphs/LightGraphs.jl] fix #1489 (#1524) 

Just add the tests and push them - let GitHub do the work for you :)

**Comment by gdalle** [comment]
There you go @sbromberger :)

**Review by sbromberger — APPROVED** [review]

**Comment by sbromberger** [comment]
Thank you, @gdalle!
